src/topology/metric_space/infsep.lean

```diff
@@ -0,0 +1,369 @@
+  { rintros ⟨x, hx, y, hy, hxy, hxyd⟩, exact ⟨(x, y), ⟨hx, hy, hxy⟩, hxyd⟩ }
+end
+
+lemma infesep_subsingleton (hs : ∀ {a b : α}, a ∈ s → b ∈ s → a = b) : s.infesep = ∞ :=
+eq_top_iff.2 (le_infesep (λ x hx y hy hxy, false.elim (hxy (hs hx hy))))
+
+@[simp] lemma infesep_empty : infesep (∅ : finset α) = ∞ :=
+infesep_subsingleton (λ _ _ ha _, false.elim (not_mem_empty _ ha))
+
+@[simp] lemma infesep_singleton : infesep ({x} : finset α) = ∞ :=
+infesep_subsingleton (λ a b ha hb, by rw [mem_singleton.1 ha, mem_singleton.1 hb])
+
+lemma infesep_anti : @antitone (finset α) _ _ _ infesep :=
+λ s t h, le_infesep $ λ x hx y hy, infesep_le_edist_of_mem (h hx) (h hy)
+
+lemma infesep_insert_le : infesep (insert x s) ≤ ⨅ (y ∈ s) (hxy : x ≠ y), edist x y :=
+begin
+  simp_rw [le_infi_iff],
+  refine λ _ hy hxy, infesep_le_edist_of_mem (mem_insert_self _ _) (mem_insert_of_mem hy) hxy
+end
+
+lemma infesep_pair_le_left (hxy : x ≠ y) : infesep ({x, y} : finset α) ≤ edist x y :=
+infesep_le_edist_of_mem (mem_insert_self _ _) (mem_insert_of_mem (mem_singleton_self _)) hxy
+
+lemma infesep_pair_le_right (hxy : x ≠ y) : infesep ({x, y} : finset α) ≤ edist y x :=
+by rw pair_comm; exact infesep_pair_le_left hxy.symm
+
+lemma le_infesep_pair : (edist x y) ⊓ (edist y x) ≤ infesep ({x, y} : finset α) :=
+begin
+  simp_rw [le_infesep_iff, inf_le_iff, mem_insert, mem_singleton],
+  rintros a (rfl | rfl) b (rfl | rfl) hab; finish
+end
+
+lemma infesep_pair_eq' (hxy : x ≠ y) : infesep ({x, y} : finset α) = (edist x y) ⊓ (edist y x) :=
+le_antisymm (_root_.le_inf (infesep_pair_le_left hxy) (infesep_pair_le_right hxy)) le_infesep_pair
+
+lemma infesep_exists (h : ∃ x y ∈ s, x ≠ y) : ∃ (x y ∈ s) (h : x ≠ y), edist x y = s.infesep :=
+begin
+  rw ← off_diag_nonempty_iff at h,
+  rcases exists_mem_eq_inf _ h (uncurry edist) with ⟨⟨x, y⟩, hxy, spec⟩,
+  rw mem_off_diag at hxy, rcases hxy with ⟨hx, hy, hxy⟩,
+  exact ⟨x, hx, y, hy, hxy, spec.symm⟩
+end
+
+lemma infesep_exists' (hx : x ∈ s) (hy : y ∈ s) (hxy : x ≠ y)
+  : ∃ (x y ∈ s) (h : x ≠ y), edist x y = s.infesep := infesep_exists ⟨x, hx, y, hy, hxy⟩
+
+/-- Classically choose a pair of distinct elements in the set which minimise the distance. -/
+noncomputable def minesep.some (hx : x ∈ s) (hy : y ∈ s) (hxy : x ≠ y) : α × α :=
+⟨(infesep_exists' hx hy hxy).some, (infesep_exists' hx hy hxy).some_spec.some_spec.some⟩
+
+lemma minesep.some_spec (hx : x ∈ s) (hy : y ∈ s) (hxy : x ≠ y) :
+(minesep.some hx hy hxy).fst ∈ s ∧ (minesep.some hx hy hxy).snd ∈ s
+∧ (minesep.some hx hy hxy).fst ≠ (minesep.some hx hy hxy).snd
+∧ edist (minesep.some hx hy hxy).fst (minesep.some hx hy hxy).snd = s.infesep :=
+⟨ (infesep_exists' hx hy hxy).some_spec.some,
+  (infesep_exists' hx hy hxy).some_spec.some_spec.some_spec.some,
+  (infesep_exists' hx hy hxy).some_spec.some_spec.some_spec.some_spec.some,
+  (infesep_exists' hx hy hxy).some_spec.some_spec.some_spec.some_spec.some_spec⟩
```

Comment:
Is this copied from a diam API? This looks very strange to me.

Reply:
No, it's complete nonsense I made up. But I put it in in order to see how it looked (bad): I'm trying to work out if you'd ever need this or if the existence is enough.

Reply:
(And I recalled Kevin saying recently that it was quite good practice to sometimes use this kind of thing.)

Reply:
A lot of the time this is just a crutch when the right tool is either `rcases infesep_exists'` or `choose ... using infesep_exists'`

Reply:
Well, that's what I'd normally do. Great, I'll strip them.